cohort-engine/src/main/java/com/ibm/cohort/engine/measure/cache/CachingRetrieveProvider.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * (C) Copyright IBM Corp. 2020, 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.cohort.engine.measure.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.opencds.cqf.cql.engine.retrieve.RetrieveProvider;
+import org.opencds.cqf.cql.engine.runtime.Code;
+import org.opencds.cqf.cql.engine.runtime.Interval;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CachingRetrieveProvider implements RetrieveProvider {
+
+	private static final Logger LOG = LoggerFactory.getLogger(CachingRetrieveProvider.class);
+
+	private final RetrieveProvider baseProvider;
+	private final Cache<CacheKey, Iterable<Object>> cache;
+
+	public CachingRetrieveProvider(RetrieveProvider baseProvider) {
+		this.baseProvider = baseProvider;
+		// TODO: Allow for configuration of the cache via...config file?...many params?
+		// What kind of configuration options to start with?  Max size.  Anything else?
+		this.cache = Caffeine.newBuilder()
+				.recordStats()
+				.maximumSize(1_000)
+				.build();
```

Comment:
I'm currently thinking through ways to extract the creation of the underlying cache into a separate class that gets passed all the way into the `CachingRetrieveProvider`.
I don't like the idea of having the initialization logic living so "deep" in the evaluation class hierarchy once the persistent cache becomes a thing.